accepted/cranelift-isel-isle-peepmatic.md

```diff
@@ -0,0 +1,926 @@
+described in this RFC, however.
+
+First, the handwritten code in our existing `MachInst`-style backends
+is fundamentally an imperative design: the machine-independent driver
+invokes a backend function once per CLIF instruction to lower, and
+this driver in turn queries the IR and eventually performs actions to
+build the lowered code (allocate temps and emit machine instructions,
+mainly). The earlier pre-RFC contrasted this approach with a
+declarative DSL one, in which the execution semantics of the DSL were
+not prescribed. Note, however, that this DSL design still takes
+efficient compilation *into account*, even if it does not prescribe
+the actual execution sequence. Specifically, the design of the
+dispatch mechanism, in which the root term (determined first by the
+entry point, then by matches on or constructions of internal terms)
+statically dispatches to a set of rules that can be combined into an
+efficient automaton, it enables a "single-pass" approach that is
+distinct from most iterative term-rewriting systems.
+
+Second, the DSL encourages a mode of thinking about instructions (both
+in the IR and at the machine level, in VCode) that is more
+value-oriented. The `LowerCtx` API is designed to be a sort of general
+query API, and instruction output registers are just attributes of an
+instruction like any other. Similarly, allocating temporary registers
+and emitting instructions are just imperative actions taken by
+lowering code; there is no notion of the emitted instructions being
+"equal" to values on the input side. In contrast, the ISLE DSL design
+privileges the notion of an instruction's "value". This makes the
+notation more natural when expressing algebraic-style reductions, and
+is consistent with many other instruction-selection systems. (As a
+particular implementation note, multiple-output instructions should
+work within this framework as well because one can treat the N-output
+tuple as a value with constructors and extractors as necessary.)
+However, ISLE does not go as far as an instruction-selector framework
+that is explicitly aware of, e.g., SSA values: rather, that is up to
+the constructors that have been defined. The ISLE system itself only
+understands terms and trees of terms; this just happens to be a
+natural abstraction with which to represent expression trees.
+
+# Discussion: Future Implications for Verification
+
+Though we have not yet worked out all the details, we are confident
+that the translation of rules expressed in the ISLE DSL into some
+machine-readable form for formal verification efforts should be
+possible. This is primarily because of the "equivalent-value"
+semantics that are inherent in a term-rewriting system. The
+denotational value of a term is the symbolic or concrete value
+produced by the instruction it represents (depending on the
+interpretation); so "all" we have to do is to write, e.g.,
+pre/post-conditions for some SMT-solver or theorem-prover that
```

Comment:
Sounds cool, do you know if someone else is doing this? Sounds like it would be a hard problem for instructions that read/write global state.

Reply:
There are various research folks working on compiler verification; e.g. [this paper](https://dl.acm.org/doi/abs/10.1145/3445814.3446751) from recent ASPLOS 2021. The general term is "translation validation" when it involves proving a specific input and output are equivalent, I think. I don't see the core Cranelift team diving into this fulltime but *enabling* it is very important, and there are academics who might be interested!

Reply:
Thanks for the paper - interesting stuff! From my brief interactions with cranelift so far, the key differences that I see with cranelift and LLVM are that neither levels of IR are defined as well in cranelift, for example:
1) I don't think that the aarch64 backend actually models pstate flag setting.
2) The CLIF icmp has some ugly target defined behaviour...
So I think having some verification would be an excellent driver for improving our internal semantics.